File: apps/apple/rishi-mcp/src/instance-registry.mjs
import { memorySnapshot } from "./memory.mjs";

export const ErrorCodes = {
  INSTANCE_NOT_FOUND: "INSTANCE_NOT_FOUND",
  INSTANCE_ALREADY_RUNNING: "INSTANCE_ALREADY_RUNNING",
  DRIVER_UNAVAILABLE: "DRIVER_UNAVAILABLE",
  ACTION_NOT_SUPPORTED: "ACTION_NOT_SUPPORTED",
  STATE_CHANGED: "STATE_CHANGED",
  WAIT_TIMEOUT: "WAIT_TIMEOUT",
};

export class RegistryError extends Error {
  constructor(code, message, data = {}) { super(message); this.code = code; this.data = data; }
}

export class InstanceRegistry {
  #driver;
  #memory;
  #instances = new Map();
  #starting = new Set();
  constructor(driver, memory = memorySnapshot) { this.#driver = driver; this.#memory = memory; }

  async list() {
    const apps = await this.#driver.listApps();
    return Promise.all(apps.filter((app) => /rishi/i.test(`${app.id} ${app.displayName ?? ""}`) && app.isRunning !== false && app.windows?.length).map(async (app) => ({ ...app, owned: [...this.#instances.values()].some((i) => i.app === app.id), memory: await this.#memory(app.id) })));
  }

  async start(app) {
    if (this.#instances.has(app) || this.#starting.has(app)) throw new RegistryError(ErrorCodes.INSTANCE_ALREADY_RUNNING, `target already owned: ${app}`, { app });
    this.#starting.add(app);
    try {
      const existing = (await this.#driver.listApps()).find((candidate) => candidate.id === app && candidate.isRunning !== false && candidate.windows?.length);
      if (existing) throw new RegistryError(ErrorCodes.INSTANCE_ALREADY_RUNNING, `target already running: ${app}`, { app, existing });
      await this.#driver.launch(app);
      const instance = { id: `${app}:${Date.now()}`, app, owned: true };
      this.#instances.set(app, instance);
      return { ...instance, memory: await this.#memory(app) };
    } finally {
      this.#starting.delete(app);
    }
  }

  async stop(app) {
    const instance = this.#instances.get(app);
    if (!instance) return { app, stopped: false, reason: "not_owned" };
    await this.#driver.terminate(app);
    this.#instances.delete(app);
    return { ...instance, stopped: true, memory: await this.#memory(app) };
  }

  async restart(app) {
    const owned = this.#instances.has(app);
    if (owned) await this.stop(app);
    else {
      const existing = (await this.#driver.listApps()).find((candidate) => candidate.id === app && candidate.isRunning !== false && candidate.windows?.length);
      if (existing) throw new RegistryError(ErrorCodes.INSTANCE_ALREADY_RUNNING, `refusing to restart unowned target: ${app}`, { app });
    }
    return this.start(app);
  }

  async cleanup() { for (const app of [...this.#instances.keys()]) await this.stop(app); }
}

File: apps/apple/rishi-mcp/Sources/RishiAppleMCP/InstanceRegistry.swift
import Foundation

public enum RegistryErrorCode: String, Sendable {
    case instanceNotFound = "INSTANCE_NOT_FOUND"
    case instanceAlreadyRunning = "INSTANCE_ALREADY_RUNNING"
    case driverUnavailable = "DRIVER_UNAVAILABLE"
    case actionNotSupported = "ACTION_NOT_SUPPORTED"
    case stateChanged = "STATE_CHANGED"
    case waitTimeout = "WAIT_TIMEOUT"
}

public struct RegistryError: Error, LocalizedError, Sendable {
    public let code: RegistryErrorCode
    public let message: String
    public let data: [String: JSONValue]
    public init(_ code: RegistryErrorCode, _ message: String, data: [String: JSONValue] = [:]) { self.code = code; self.message = message; self.data = data }
    public var errorDescription: String? { message }
}

public struct AppInstance: Sendable, Equatable {
    public let id: String
    public let displayName: String
    public let isRunning: Bool
    public let windowCount: Int
    public init(id: String, displayName: String, isRunning: Bool, windowCount: Int) { self.id = id; self.displayName = displayName; self.isRunning = isRunning; self.windowCount = windowCount }
    public var json: JSONValue { .object(["id": .string(id), "displayName": .string(displayName), "isRunning": .bool(isRunning), "windows": .array(Array(repeating: .object(["id": .integer(1), "app": .string(id)]), count: windowCount))]) }
}

public protocol AppleAppDriver: Sendable {
    func listApps() async throws -> [AppInstance]
    func launch(_ target: String) async throws
    func terminate(_ target: String) async throws
    func state(_ target: String, screenshot: Bool) async throws -> JSONValue
    func logs(_ target: String, limit: Int) async throws -> JSONValue
    func request(_ target: String, payload: JSONValue, timeoutMs: Int) async throws -> JSONValue
    func clickIdentifier(_ target: String, identifier: String, action: String) async throws -> JSONValue
    func clickText(_ target: String, text: String) async throws -> JSONValue
    func openURL(_ target: String, url: String) async throws -> JSONValue
}

public protocol MemorySnapshotting: Sendable {
    func snapshot(match: String) async throws -> JSONValue
}

public actor InstanceRegistry {
    private let driver: any AppleAppDriver
    private let memory: any MemorySnapshotting
    private var instances: [String: JSONValue] = [:]
    private var starting: Set<String> = []

    public init(driver: any AppleAppDriver, memory: any MemorySnapshotting) { self.driver = driver; self.memory = memory }

    public func list() async throws -> JSONValue {
        let apps = try await driver.listApps()
        var result: [JSONValue] = []
        for app in apps where app.id.range(of: "rishi", options: .caseInsensitive) != nil && app.isRunning && app.windowCount > 0 {
            var value = app.json.objectValue ?? [:]
            value["owned"] = .bool(instances.values.contains { $0["app"]?.stringValue == app.id })
            value["memory"] = try await memory.snapshot(match: app.id)
            result.append(.object(value))
        }
        return .array(result)
    }

    public func start(_ app: String) async throws -> JSONValue {
        guard !instances.keys.contains(app), !starting.contains(app) else { throw RegistryError(.instanceAlreadyRunning, "target already owned: \(app)", data: ["app": .string(app)]) }
        starting.insert(app)
        defer { starting.remove(app) }
        if let existing = try await driver.listApps().first(where: { $0.id == app && $0.isRunning && $0.windowCount > 0 }) {
            throw RegistryError(.instanceAlreadyRunning, "target already running: \(app)", data: ["app": .string(app), "existing": existing.json])
        }
        try await driver.launch(app)
        let value: JSONValue
        do {
            value = .object(["id": .string("\(app):\(Int(Date().timeIntervalSince1970 * 1000))"), "app": .string(app), "owned": .bool(true), "memory": try await memory.snapshot(match: app)])
        } catch {
            // The driver may have launched a real XCTest/app process before
            // the initial snapshot failed. Do not lose ownership of that
            // process just because the bookkeeping response could not be
            // produced.
            try? await driver.terminate(app)
            throw error
        }
        instances[app] = value
        return value
    }

    public func stop(_ app: String) async throws -> JSONValue {
        guard let instance = instances[app] else { return .object(["app": .string(app), "stopped": .bool(false), "reason": .string("not_owned")]) }
        try await driver.terminate(app)
        instances.removeValue(forKey: app)
        var result = instance.objectValue ?? [:]
        result["stopped"] = .bool(true)
        result["memory"] = try await memory.snapshot(match: app)
        return .object(result)
    }

    public func restart(_ app: String) async throws -> JSONValue {
        if instances[app] != nil { _ = try await stop(app) }
        else if let existing = try await driver.listApps().first(where: { $0.id == app && $0.isRunning && $0.windowCount > 0 }) {
            throw RegistryError(.instanceAlreadyRunning, "refusing to restart unowned target: \(app)", data: ["app": .string(app), "existing": existing.json])
        }
        return try await start(app)
    }

    public func cleanup() async throws {
        var firstError: Error?
        for app in Array(instances.keys) {
            do { _ = try await stop(app) }
            catch { if firstError == nil { firstError = error } }
        }
        if let firstError { throw firstError }
    }
}
